### ISSUE TITLE
operator/tests: Each operator test should run with a fresh instance of operator 

### ISSUE THREAD

**Issue by avalluri** [intel/pmem-csi#593]
A new operator instance is supposed to run before each test run and should get deleted after the test. This ensures that the operator is not using any cache generated by the previous test.

Below logs shows that it is using previously running operator instance:
```
[
2020-04-18T03:24:49.922Z] [90m------------------------------[0m
[2020-04-18T03:24:49.922Z] [0moperator[0m [90mAPI[0m [0mdeployment[0m 
[2020-04-18T03:24:49.922Z]   [1mshall not allow multiple deployments with same driver name[0m
[2020-04-18T03:24:49.922Z]   [37m/mnt/workspace/pmem-csi_PR-584/test/e2e/operator/deployment_api.go:419[0m
[2020-04-18T03:24:49.922Z] [BeforeEach] operator
[2020-04-18T03:24:49.922Z]   /mnt/workspace/gopath/pkg/mod/github.com/pohly/kubernetes@v1.18.1-pmem-csi-20200416/test/e2e/framework/framework.go:178
[2020-04-18T03:24:49.922Z] [1mSTEP[0m: Creating a kubernetes client
[2020-04-18T03:24:49.922Z] Apr 18 03:24:49.495: INFO: >>> kubeConfig: /mnt/workspace/pmem-csi_PR-584/_work/govm/kube.config
[2020-04-18T03:24:49.922Z] [BeforeEach] operator
[2020-04-18T03:24:49.922Z]   /mnt/workspace/pmem-csi_PR-584/test/e2e/deploy/deploy.go:533
[2020-04-18T03:24:49.922Z] [1mSTEP[0m: preparing for test "operator API deployment shall not allow multiple deployments with same driver name"
[2020-04-18T03:24:49.922Z] Apr 18 03:24:49.515: INFO: No external IP address on nodes, falling back to internal IPs
[2020-04-18T03:24:49.922Z] Apr 18 03:24:49.579: INFO: reusing existing operator PMEM-CSI components
[2020-04-18T03:24:49.922Z] [1mSTEP[0m: Operator is ready!
```

**Comment by avalluri** [comment]
I believe this test failure also related to this issue: 
https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/blue/rest/organizations/jenkins/pipelines/pmem-csi/branches/PR-584/runs/25/nodes/88/log/?start=0

On beggening of test it tries to start the operator, but it found existing operator and so it tries to use :
```
[2020-04-18T09:11:23.058Z] [0moperator[0m [90mAPI[0m [0mdeployment[0m 
[2020-04-18T09:11:23.058Z]   [1mshould be able to capture deployment changes operator when not running[0m
[2020-04-18T09:11:23.058Z]   [37m/mnt/workspace/pmem-csi_PR-584/test/e2e/operator/deployment_api.go:554[0m
[2020-04-18T09:11:23.058Z] [BeforeEach] operator
[2020-04-18T09:11:23.058Z]   /mnt/workspace/gopath/pkg/mod/github.com/pohly/kubernetes@v1.18.1-pmem-csi-20200416/test/e2e/framework/framework.go:178
[2020-04-18T09:11:23.058Z] [1mSTEP[0m: Creating a kubernetes client
[2020-04-18T09:11:23.058Z] Apr 18 09:11:22.593: INFO: >>> kubeConfig: /mnt/workspace/pmem-csi_PR-584/_work/govm/kube.config
[2020-04-18T09:11:23.058Z] [BeforeEach] operator
[2020-04-18T09:11:23.058Z]   /mnt/workspace/pmem-csi_PR-584/test/e2e/deploy/deploy.go:533
[2020-04-18T09:11:23.058Z] [1mSTEP[0m: preparing for test "operator API deployment should be able to capture deployment changes operator when not running"
[2020-04-18T09:11:23.058Z] Apr 18 09:11:22.604: INFO: No external IP address on nodes, falling back to internal IPs
[2020-04-18T09:11:23.058Z] Apr 18 09:11:22.627: INFO: reusing existing operator PMEM-CSI components
```

But fails with timeout:
```
020-04-18T09:16:26.993Z] [91m[1m• Failure in Spec Setup (BeforeEach) [300.039 seconds][0m
[2020-04-18T09:16:26.993Z] operator
[2020-04-18T09:16:26.993Z] [90m/mnt/workspace/pmem-csi_PR-584/test/e2e/deploy/deploy.go:708[0m
[2020-04-18T09:16:26.993Z]   [91m[1mAPI [BeforeEach][0m
[2020-04-18T09:16:26.993Z]   [90m/mnt/workspace/pmem-csi_PR-584/test/e2e/deploy/deploy.go:695[0m
[2020-04-18T09:16:26.993Z]     deployment
[2020-04-18T09:16:26.993Z]     [90m/mnt/workspace/pmem-csi_PR-584/test/e2e/operator/deployment_api.go:39[0m
[2020-04-18T09:16:26.993Z]       should be able to capture deployment changes operator when not running
[2020-04-18T09:16:26.993Z]       [90m/mnt/workspace/pmem-csi_PR-584/test/e2e/operator/deployment_api.go:554[0m
[2020-04-18T09:16:26.993Z] 
[2020-04-18T09:16:26.993Z]       [91mTimed out after 300.000s.
[2020-04-18T09:16:26.993Z]       operator not running in namespace default
[2020-04-18T09:16:26.993Z]       Expected
[2020-04-18T09:16:26.993Z]           <bool>: false
[2020-04-18T09:16:26.993Z]       to be true[0m
[2020-04-18T09:16:26.993Z] 
[2020-04-18T09:16:26.993Z]       /mnt/workspace/pmem-csi_PR-584/test/e2e/deploy/deploy.go:69
[2020-04-18T09:16:26.993Z] [90m------------------------------[0m
```

**Comment by pohly** [comment]
> A new operator instance is supposed to run before each test run and should get deleted after the test. This ensures that the operator is not using any cache generated by the previous test.

One can also argue the opposite: in practice, the operator is only installed once in a cluster and then needs to deploy and remove drivers many times. If that fails because of cached data, then testing has found a real problem.

> On beggening of test it tries to start the operator, but it found existing operator and so it tries to use :
...
> The reason was the previously running test stops the operator to ensure that the existing drivers still running

Two solutions:
- `EnsureDeployment` needs to be improved such that it fixes incomplete deployments (useful either way).
- We add a new pseudo deployment named "none" which removes any running driver and operator before starting tests, and tests which want control over the running operator use that.

**Comment by pohly** [comment]
> EnsureDeployment needs to be improved such that it fixes incomplete deployments (useful either way).

Here's an implementation of that approach. It applies on top of your current PR #584. The `dep.GetName()` is an unrelated simplification that occurred to me after having learned about the `runtime.Runtime` API.

```
diff --git a/test/e2e/deploy/deploy.go b/test/e2e/deploy/deploy.go
index f301c252..86b81883 100644
--- a/test/e2e/deploy/deploy.go
+++ b/test/e2e/deploy/deploy.go
@@ -539,20 +539,15 @@ func EnsureDeployment(deploymentName string) *Deployment {
 		if running != nil {
 			if reflect.DeepEqual(deployment, running) {
 				framework.Logf("reusing existing %s PMEM-CSI components", deployment.Name)
-				// Do some sanity checks on the running deployment before the test.
-				if deployment.HasDriver {
-					WaitForPMEMDriver(c, deployment.Namespace)
-				}
-				if deployment.HasOperator {
-					WaitForOperator(c, deployment.Namespace)
-				}
-				return
+			} else {
+				framework.Logf("have %s PMEM-CSI deployment, want %s -> delete existing deployment", running.Name, deployment.Name)
+				err := RemoveObjects(c, running.Name)
+				framework.ExpectNoError(err, "remove PMEM-CSI deployment")
 			}
-			framework.Logf("have %s PMEM-CSI deployment, want %s -> delete existing deployment", running.Name, deployment.Name)
-			err := RemoveObjects(c, running.Name)
-			framework.ExpectNoError(err, "remove PMEM-CSI deployment")
 		}
 
+		// The code below runs also for existing deployments. It must be idempotent and
+		// fix the existing deployment in case that it was modified by the previous test.
 		if deployment.HasOperator {
 			// At the moment, the only supported deployment method is via test/start-operator.sh.
 			cmd := exec.Command("test/start-operator.sh")
@@ -600,7 +595,7 @@ func EnsureDeployment(deploymentName string) *Deployment {
 				depUnstructured := &unstructured.Unstructured{
 					Object: hash,
 				}
-				CreateDeploymentCR(f, depUnstructured)
+				EnsureDeploymentCR(f, depUnstructured)
 			} else {
 				// Deploy with script.
 				cmd := exec.Command("test/setup-deployment.sh")
diff --git a/test/e2e/deploy/operator.go b/test/e2e/deploy/operator.go
index 17e7fd63..c4d663d1 100644
--- a/test/e2e/deploy/operator.go
+++ b/test/e2e/deploy/operator.go
@@ -43,8 +43,7 @@ func init() {
 
 func CreateDeploymentCR(f *framework.Framework, dep *unstructured.Unstructured) *unstructured.Unstructured {
 	var outDep *unstructured.Unstructured
-	metadata := dep.Object["metadata"].(map[string]interface{})
-	depName := metadata["name"].(string)
+	depName := dep.GetName()
 	gomega.Eventually(func() error {
 		var err error
 		outDep, err = f.DynamicClient.Resource(DeploymentResource).Create(context.Background(), dep, metav1.CreateOptions{})
@@ -55,6 +54,28 @@ func CreateDeploymentCR(f *framework.Framework, dep *unstructured.Unstructured)
 	return outDep
 }
 
+func EnsureDeploymentCR(f *framework.Framework, dep *unstructured.Unstructured) *unstructured.Unstructured {
+	var outDep *unstructured.Unstructured
+	depName := dep.GetName()
+	gomega.Eventually(func() error {
+		existingDep, err := f.DynamicClient.Resource(DeploymentResource).Get(context.Background(), depName, metav1.GetOptions{})
+		if err == nil {
+			dep.SetResourceVersion(existingDep.GetResourceVersion())
+			outDep, err = f.DynamicClient.Resource(DeploymentResource).Update(context.Background(), dep, metav1.UpdateOptions{})
+			LogError(err, "update deployment error: %v, will retry...", err)
+			return err
+		}
+		if apierrs.IsNotFound(err) {
+			outDep, err = f.DynamicClient.Resource(DeploymentResource).Create(context.Background(), dep, metav1.CreateOptions{})
+			LogError(err, "create deployment error: %v, will retry...", err)
+			return err
+		}
+		return err
+	}, "3m", "10s").Should(gomega.BeNil(), "create deployment %q", depName)
+	ginkgo.By(fmt.Sprintf("Created deployment %q", depName))
+	return outDep
+}
+
 func DeleteDeploymentCR(f *framework.Framework, name string) {
 	gomega.Eventually(func() error {
 		err := f.DynamicClient.Resource(DeploymentResource).Delete(context.Background(), name, metav1.DeleteOptions{})
@@ -69,8 +90,7 @@ func DeleteDeploymentCR(f *framework.Framework, name string) {
 
 func UpdateDeploymentCR(f *framework.Framework, dep *unstructured.Unstructured) *unstructured.Unstructured {
 	var outDep *unstructured.Unstructured
-	metadata := dep.Object["metadata"].(map[string]interface{})
-	depName := metadata["name"].(string)
+	depName := dep.GetName()
 
 	gomega.Eventually(func() error {
 		var err error
diff --git a/test/e2e/operator/deployment_api.go b/test/e2e/operator/deployment_api.go
index 9a7c8595..0058d455 100644
--- a/test/e2e/operator/deployment_api.go
+++ b/test/e2e/operator/deployment_api.go
@@ -37,12 +37,6 @@ var _ = deploy.DescribeForSome("API", func(d *deploy.Deployment) bool {
 
 	f := framework.NewDefaultFramework("API")
 	f.SkipNamespaceCreation = true
-	f.AddAfterEach("API", func(f *framework.Framework, _ bool) {
-		framework.Logf("Removing PMEM-CSI operator '%s' in '%s' namespace", d.Name, d.Namespace)
-		Expect(f).ShouldNot(BeNil(), "framework initialization")
-		err := deploy.RemoveObjects(c, d.Name)
-		framework.ExpectNoError(err, "remove PMEM-CSI deployment")
-	})
 
 	BeforeEach(func() {
 		var err error
```

**Comment by pohly** [comment]
I've submitted that enhancement in https://github.com/intel/pmem-csi/pull/600.

**Comment by avalluri** [comment]
> One can also argue the opposite: in practice, the operator is only installed once in a cluster and then needs to deploy and remove drivers many times. If that fails because of cached data, then testing has found a real problem.

This could be/already covered by a test specifically that deploys multiple deployments and ensures that everything works as expected. Otherwise, running a single operator for all tests might give false positives which are not really errors.

**Comment by avalluri** [comment]
> We add a new pseudo deployment named "none" which removes any running driver and operator before starting tests, and tests which want control over the running operator use that.

As the current `test/e2e/deploy.go` has the infrastructure to start operator if not running, the operator tests could remove the operator after each test. This ensures that a new operator is deployed for each tests. I implemented this here: 
https://github.com/intel/pmem-csi/pull/584/commits/97ddf69fa44c8c148b50d686972d39bd58218a82

**Comment by pohly** [comment]
> running a single operator for all tests might give false positives which are not really errors

Why? Can you give an example?

> This ensures that a new operator is deployed for each tests.

We could do that. The question is: should we?

I don't think we should. Our default mode of operation should be as close as possible to real usage, to uncover bugs in that mode. And that mode is "deploy operator once, do different things with it".

We do the same with driver installation (deploy once, run different tests) and (ultimately) with the cluster itself. Why is the testing of the operator different?

**Comment by pohly** [comment]
> BTW, Shouldn't we stop the operator/driver at the end of running the test suite just to make sure to restore the clean state?

That would slow down running the same test repeatedly with "make test_e2e TEST_E2E_FOCUS=<some test>". I prefer to keep things running.

The only reliable way to get a clean slate is to stop the cluster and restart it. Stopping the driver doesn't ensure that other garbage like leaked volumes get cleaned up.
